Documents/UpgradeExtensibility.md

```diff
@@ -0,0 +1,157 @@
+VFSForGit Upgrade logic, as it would only need to run a single
+installer, but the installer generation would be more complex, as it
+would need to include logic for all updated packaged - including the
+git and VFSForGit installers. As we already have the logic to run the
+git and VFSForGit installers, we decided against this option because
+it adds more complexity to the installer generation.
+
+#### Alternative 2
+
+Git and GVFS would be upgraded from the public feed as normal, and
+only the extra components would be installed from the specified
+feed. We decided against this because the components to be distributed
+can be coupled, and having two independent feeds could make it
+difficult to keep these in sync.
+
+### Upgrade Logic
+
+There are currently two high level components in VFSForGit:
+
+UpgradeOrchestrator :: Responsible for orchestrating the upgrade steps
+
+ProductUpgrader :: Responsible for details of interacting with GitHub
+Releases API, downloading assets, and performing the upgrade of
+individual assets.
+
+We will make the responsibilities of these components more clear, such
+that the UpgradeOrchestrator will be responsible for the high level
+general steps of an upgrade, while the ProductUpgrader will contain
+the logic for interacting with the different feeds and installing
+packages from these feeds. In this model, we would have a
+"GitHubReleasesUpgrader" and a "NuGetPackageUpgrader"
+
+The UpgradeOrchestrator will delegate the actual upgrade logic to
+individual ProductUpgraders. The existing upgrade logic for upgrading
+from the public GitHub releases API will be one ProductUpgrader
+implementation. We will implement another ProductUpgrader for the
+upgrades from NuGet feeds.
+
+- Checking for a new version
+  - Delegate the logic of querying the feed to Upgrader
+- Running prechecks
+- Downloading new version
+  - Download logic delegated to ProductUpgrader
+- Unmounting all repositories
+- Running installer
+  - Installer logic delegated to ProductUpgrader
+- Remounting repositories
+- Deleting downloaded files
+  - cleanup delegated to Upgrader
+
+#### Upgrade with Alternative feed
+
+When upgrading from NuGet package feed, the NuGet package will be
+downloaded and the individual well know directories extracted. The
+installers from these directories will then be run.
+
+### Checking for new upgrades
+
+ProductUpgrader currently contains the logic for determining when an
+upgrade is available. It currently does this by checking if an
+installer has been downloaded.
+
+Now that we have to handle different feed sources, the installers will
+need to follow a convention for us to read the version from, or we
+might want to put the version infromation in a well known file. The
+advantage of storing the version information in a file is that we can
+update this file by querying the feed endpoints, and not necessarily
+waiting to download the actual installer.
+
+### TBD:

```

Comment:
I think it's worth calling out any scenarios that you don't think we should support (or if we should support them, include the details of how they will work).  

Some scenarios I'm not sure about:

- Switching to\from alternate feeds to public feeds (if a user switches to a public feed, will we run an uninstaller for the custom components)?
- Switching to\from feeds in different orgs
- Switching to\from different feeds in the same org (I assume we do want to support this)